examples/graphbolt/disk_based_feature/node_classification_offline.py

```diff
@@ -0,0 +1,644 @@
+    of sampled nodes in a mini-batch, otherwise the feature fetcher might get into
+    a deadlock, causing a hang.
+    """
+    if args.cpu_cache_size_in_gigabytes > 0 and isinstance(
+        features[("node", None, "feat")], gb.DiskBasedFeature
+    ):
+        features[("node", None, "feat")] = gb.CPUCachedFeature(
+            features[("node", None, "feat")],
+            int(args.cpu_cache_size_in_gigabytes * 1000 * 1000 * 1000),
+            args.cpu_feature_cache_policy,
+            args.feature_device == "pinned",
+        )
+        cpu_cached_feature = features[("node", None, "feat")]
+        cpu_cache_miss_rate_fn = lambda: cpu_cached_feature._feature.miss_rate
+    else:
+        cpu_cache_miss_rate_fn = lambda: 1
+
+    """
+    If the GPU cache size is greater than 0, we wrap the underlying feature store
+    to be a GPUCachedFeature. This will reduce the amount of data transferred during
+    host-to-device copy operations for this feature.
+    """
+    if args.gpu_cache_size_in_gigabytes > 0 and args.feature_device != "cuda":
+        features[("node", None, "feat")] = gb.GPUCachedFeature(
+            features[("node", None, "feat")],
+            int(args.gpu_cache_size_in_gigabytes * 1000 * 1000 * 1000),
+        )
+        gpu_cached_feature = features[("node", None, "feat")]
+        gpu_cache_miss_rate_fn = lambda: gpu_cached_feature._feature.miss_rate
+    else:
+        gpu_cache_miss_rate_fn = lambda: 1
+
+    num_minibatch = (len(train_set) + args.batch_size - 1) // args.batch_size
+    train_sampler = create_sampler(
+        graph=graph,
+        itemset=train_set,
+        batch_size=args.batch_size,
+        fanout=args.fanout,
+        device=args.device,
+        job="train",
+    )
+    train_dataloader = create_trainloader(
+        features=features,
+        itemset=gb.ItemSet(num_minibatch, "seeds"),
+        device=args.device,
+        job="train",
+    )
+    valid_dataloader = create_dataloader(
+        graph=graph,
+        features=features,
+        itemset=valid_set,
+        batch_size=args.batch_size,
+        fanout=args.fanout,
+        device=args.device,
+        job="evaluate",
+    )
+
+    in_channels = features.size("node", None, "feat")[0]
+    model = SAGE(
+        in_channels,
+        args.num_hidden,
+        num_classes,
+        len(args.fanout),
+        args.dropout,
+    ).to(args.device)
+    assert len(args.fanout) == len(model.layers)
+
+    perpare_time = time.time() - start
+    print(f"Prepare time: {perpare_time:.2f}s")
+
+    start = time.time()
+    if not os.path.exists(args.subgraph_dir):
+        os.makedirs(args.subgraph_dir)
+        for it, minibatch in enumerate(tqdm(train_sampler, "Sampling")):
+            data = [
+                minibatch.seeds.cpu(),
+                minibatch.input_nodes.cpu(),
+                minibatch.labels.cpu(),
+                [block.cpu() for block in minibatch.blocks],
```

Comment:
This will only work for DGL. I would recommend changing your code so that it works with any GNN framework. Otherwise, your methods should be called DGLMinibatchProvider or something similar.

Reply:
For now this is just for alignment with DiskGNN, as if I directly save all attributes in a minibatch, the size will be large and loading the minibatches is time consuming.

Reply:
You don't have to save all attributes in minibatch, you can only save `minibatch.sampled_subgraphs` as it is the counterpart to DGL's blocks.